mapic/utils.go

```diff
@@ -0,0 +1,58 @@
+package mistapiconnector
+
+import (
+	"github.com/livepeer/go-api-client"
+	"sync"
+	"time"
+)
+
+const lapiCacheDuration = 1 * time.Second
+
+type ApiClientCached struct {
+	lapi        *api.Client
+	streamCache map[string]entry
+	mu          sync.RWMutex
+	ttl         time.Duration
+}
+
+func NewApiClientCached(lapi *api.Client) *ApiClientCached {
+	return &ApiClientCached{
+		lapi:        lapi,
+		streamCache: make(map[string]entry),
+		ttl:         lapiCacheDuration,
+	}
+}
+
+type entry struct {
+	stream   *api.Stream
+	err      error
+	updateAt time.Time
+}
+
+func (a *ApiClientCached) GetStreamByPlaybackID(playbackId string) (*api.Stream, error) {
+	a.mu.RLock()
+	e, ok := a.streamCache[playbackId]
+	if ok && e.stream != nil && e.updateAt.Add(a.ttl).After(time.Now()) {
+		// Use cached value
+		a.mu.RUnlock()
+		return e.stream, e.err
+	}
+	a.mu.RUnlock()
+
+	// Value not cached or expired
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	// Check again in case another goroutine has updated the cache in the meantime
+	e, ok = a.streamCache[playbackId]
+	if ok && e.stream != nil && e.updateAt.Add(a.ttl).After(time.Now()) {
+		return e.stream, e.err
+	}
+	// No value in the cache, fetch from Livepeer API and store the result in a cache
+	stream, err := a.lapi.GetStreamByPlaybackID(playbackId)
+	a.streamCache[playbackId] = entry{
+		stream:   stream,
+		err:      err,
```

Comment:
do we need to cache error cases if we only use cache when it's successful? Could be clearer to have an early return for `err != nil` here instead and skip the err in cache

Reply:
We do want to cache errors as well. That was actually the root cause of the issue that we never cached failed requests so they piled up.